ar71xx: add support for Gainstrong MiniBox V3.2

p4u opened the pull request:
Adds MiniBox V3.2 ar71xx target (QCA99531 + QCA9887)
Info: http://oolite.cn/minibox-v3-2-qca9531-qca9887-module-pin-specification.html
Reference patch: http://oolite.cn/wp-content/uploads/2017/09/minibox_v3_cc15.05_20160325.patch
Installation: Upload firmware to uboot via HTTP at 192.168.1.1 (press button for 5 seconds)

Signed-off-by: Pau Escrich <p4u@dabax.net>

p4u committed:
ar71xx: add support for Gainstrong MiniBox V3.2

Adds MiniBox V3.2 ar71xx target (QCA99531 + QCA9887)
Info: http://oolite.cn/minibox-v3-2-qca9531-qca9887-module-pin-specification.html
Reference patch: http://oolite.cn/wp-content/uploads/2017/09/minibox_v3_cc15.05_20160325.patch
Installation: Upload firmware to uboot via HTTP at 192.168.1.1 (press button for 5 seconds)

Signed-off-by: Pau Escrich <p4u@dabax.net>

p4u commented:
Updates https://github.com/lede-project/source/pull/1622

blogic commented on target/linux/ar71xx/base-files/lib/ar71xx.sh:
```diff
@@ -287,11 +287,16 @@ tplink_board_detect() {
 		model="OOLITE"
+		[ "$hwid" = "3c00010c" -a "$mid" = "00000001" ] && {
+			model="MINIBOX_V3.2"
+			hwver=""
```
set it to v3.2

danielg4 replied:
This was my fault. I supplied these lines when responding to a request for help programmatically differentiating this device from the Oolite and the MiniBox v1. On second glance, it appears this code may not be even called, because MIPS_MACHINE does not yield a string having "TP-LINK " in front of it. IIRC, it had some use for the OEM firmware's flashing function.

blogic commented on target/linux/ar71xx/base-files/lib/ar71xx.sh:
```diff
@@ -287,11 +287,16 @@ tplink_board_detect() {
+	"3c0002"*)
 		model="MINIBOX_V1"
+		hwver=""
 		;;
```
and v1 here

danielg4 replied:
This line also.

blogic commented on target/linux/ar71xx/files/arch/mips/ath79/mach-gs-minibox-v32.c:
```diff
@@ -0,0 +1,120 @@
+static const char *gs_minibox_v3_part_probes[] = {
+  "tp-link",
+  NULL,
+};
```
this whole file uses spaces and not tabs for indenting. please use tabs

p4u commented:
I fixed the requested changes. But now there is a conflict in the merging. Kernel config files has changed in openwrt/master branch.

Is there an elegant way to solve this without recreating the whole patch?

danielg4 commented:
Have you tried "git rebase" to restage your changes onto the current master?

p4u commented:
I've tried

> git remote add official <openwrt_source_url>
> git fetch official
> git merge official/master
> -> solve conflicts and merge
> git rebase official/master
> -> big mess of conflicts again, no way to have a clean single commit from my PR
>

p4u commented:
Great, rebase and fix conflicts worked!

Now it's ready to be merged.

p4u commented:
ping @blogic

p4u commented:
Wait! i've compiled this last version and it does not work

Find attach the serial console output.

I'll look at this afterwars, but any help is welcome.

[minibox_output.txt](https://github.com/openwrt/openwrt/files/1686611/minibox_output.txt)

blogic commented:
@p4u enable the symbol inside target/linux/ar71xx/generic/config-default ... your mach file is currently not being built into the kernel

p4u commented:
Thanks @blogic the kernel config was lost after the rebase with OpenWRT/master branch

I've updated the PR. I've just tested and works as expected. So it might be merged if no other changes requested.

Thanks.

p4u commented:
I've rebased and fixed conflicts.

However I don't know why the PR is not getting updated (at least until now).

Here the new version of the commit

https://github.com/p4u/openwrt/commit/3fb37b42bec2592f2cbbc1211989dfcd91b3c841

I'm getting a bit tired of rebasing... this might be merged already.

p4u commented:
@pepe2k made some commits chanching the behaviour of Gainstrong device

https://github.com/openwrt/openwrt/commit/907d67491068890e27f7611da5c5c4d44a672c3c#diff-f2e8b3d101abc0946e31748fec7f17b0

I'm not sure how to proceed. @pepe2k can you make the proper modifications to include also miniBox v3.2 base on the work I did or give me the instructions for recreating the PR.

I'd like to see minibox v3.2 supported as soon as possible, it's a great device.

Thanks

aparcar commented:
@p4u @blogic hey how to proceed regarding this issue?

aparcar commented:
https://git.openwrt.org/?p=openwrt/staging/dangole.git;a=commit;h=bdafdabaa1fd560cba1824fe5c5b863da0f12c26

p4u commented:
This patch fixes de problem.

```
diff --git a/target/linux/ar71xx/generic/config-default b/target/linux/ar71xx/generic/config-default
index da331d3..e5cee89 100644
--- a/target/linux/ar71xx/generic/config-default
+++ b/target/linux/ar71xx/generic/config-default
@@ -103,6 +103,7 @@ CONFIG_ATH79_MACH_GL_MIFI=y
 CONFIG_ATH79_MACH_GL_USB150=y
 CONFIG_ATH79_MACH_GS_OOLITE_V1=y
 CONFIG_ATH79_MACH_GS_OOLITE_V5_2=y
+CONFIG_ATH79_MACH_GS_MINIBOX_V32=y
 CONFIG_ATH79_MACH_HIWIFI_HC6361=y
 CONFIG_ATH79_MACH_HORNET_UB=y
 CONFIG_ATH79_MACH_JA76PF=y
```

dangowrt commented:
> This patch fixes de problem.
> 
> ```
> diff --git a/target/linux/ar71xx/generic/config-default b/target/linux/ar71xx/generic/config-default
> index da331d3..e5cee89 100644
> --- a/target/linux/ar71xx/generic/config-default
> +++ b/target/linux/ar71xx/generic/config-default
> @@ -103,6 +103,7 @@ CONFIG_ATH79_MACH_GL_MIFI=y
>  CONFIG_ATH79_MACH_GL_USB150=y
>  CONFIG_ATH79_MACH_GS_OOLITE_V1=y
>  CONFIG_ATH79_MACH_GS_OOLITE_V5_2=y
> +CONFIG_ATH79_MACH_GS_MINIBOX_V32=y
>  CONFIG_ATH79_MACH_HIWIFI_HC6361=y
>  CONFIG_ATH79_MACH_HORNET_UB=y
>  CONFIG_ATH79_MACH_JA76PF=y
> ```

Done. See https://git.openwrt.org/?p=openwrt/staging/dangole.git;a=blobdiff;f=target/linux/ar71xx/generic/config-default;h=25b58ae91e9919fce6f63f9b8fb2adef8239ef82;hp=26717b107c75c1c0ecb622140339446a1566ac97;hb=8c3f669172b97dfc738e1f6cc22bcbe161c5015a;hpb=860de2e1aa9deac7dce76b7f85cb5fcfbba00468

Please re-test and report back.

p4u commented:
Everything works quite fine now. The only proplem I found is the serial console, I can see Uboot and the begining of OpenWRT kernel loading but afterwars I loose the output. Probably the Serial device is wrong, I think it was fixed somewhere in the old patches.  I'll take a look afterwars.

Anyway it's probably good enough to be merged.

p4u commented:
By the way, two of the four devices I'm using for testing have no MAC address in the 5GHz radio. Using "option macaddr blah" in /etc/config/wireless seem to work but the WiFi device is not working properly. 

I'd like to think this is just a corner case and I got bad luck when buying the devices (from a wrong batch). Not sure if something can be done from the OS level.

p4u commented:
Regarding the serial console issue:

In the filesystem I see /dev/ttyS0 and it was changed from ttyATH0 to ttyS0 following @pepe2k recommendation (see https://github.com/lede-project/source/pull/1578)

So I think the serial device is actually ttyS0 (if I don't remember wrong it used to work). I'll try now a firmware with the CONSOLE line removed (as it is in Oolite v5.2 which is almost the same hardware) and report back.

p4u commented:
By the way, which ath10 driver should be used? The standard or `-ct` one? Should I run some tests to figure out?

dangowrt commented:
According to the bootlog the console should indeed by `ttyS0,115200` which is the default and we should just drop that line entirely. Can you try that and report whether that fixes it?

p4u commented:
> According to the bootlog the console should indeed by `ttyS0,115200` which is the default and we should just drop that line entirely. Can you try that and report whether that fixes it?

Confirmed, removing the CONSOLE line fixes de problem :)

p4u commented:
So, everything works good now. Just the EEPROM missing mac problem in some devices, but I think this is not an OpenWRT issue. Just in case it might be useful, the logread information regarding ath10k driver.

```
Mon Nov  5 21:24:19 2018 kern.info kernel: [   12.197625] ath10k_pci 0000:00:00.0: pci irq legacy oper_irq_mode 1 irq_mode 0 reset_mode 0
Mon Nov  5 21:24:19 2018 kern.warn kernel: [   12.479380] ath10k_pci 0000:00:00.0: Direct firmware load for ath10k/pre-cal-pci-0000:00:00.0.bin failed with error -2
Mon Nov  5 21:24:19 2018 kern.warn kernel: [   12.490503] ath10k_pci 0000:00:00.0: Falling back to user helper
Mon Nov  5 21:24:19 2018 kern.err kernel: [   12.592057] firmware ath10k!pre-cal-pci-0000:00:00.0.bin: firmware_loading_store: map pages failed
Mon Nov  5 21:24:19 2018 kern.warn kernel: [   12.601845] ath10k_pci 0000:00:00.0: Direct firmware load for ath10k/cal-pci-0000:00:00.0.bin failed with error -2
Mon Nov  5 21:24:19 2018 kern.warn kernel: [   12.612601] ath10k_pci 0000:00:00.0: Falling back to user helper
Mon Nov  5 21:24:19 2018 kern.err kernel: [   12.806212] firmware ath10k!cal-pci-0000:00:00.0.bin: firmware_loading_store: map pages failed
Mon Nov  5 21:24:19 2018 kern.warn kernel: [   12.815666] ath10k_pci 0000:00:00.0: Direct firmware load for ath10k/QCA9887/hw1.0/firmware-6.bin failed with error -2
Mon Nov  5 21:24:19 2018 kern.warn kernel: [   12.826781] ath10k_pci 0000:00:00.0: Falling back to user helper
Mon Nov  5 21:24:19 2018 kern.err kernel: [   13.020657] firmware ath10k!QCA9887!hw1.0!firmware-6.bin: firmware_loading_store: map pages failed
Mon Nov  5 21:24:19 2018 kern.info kernel: [   13.040419] ath10k_pci 0000:00:00.0: qca9887 hw1.0 target 0x4100016d chip_id 0x004000ff sub 0000:0000
Mon Nov  5 21:24:19 2018 kern.info kernel: [   13.050055] ath10k_pci 0000:00:00.0: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 1
Mon Nov  5 21:24:19 2018 kern.info kernel: [   13.065217] ath10k_pci 0000:00:00.0: firmware ver 10.2.4-1.0-00041 api 5 features no-p2p,ignore-otp,skip-clock-init,mfp,allows-mesh-bcast crc32 3587a422
Mon Nov  5 21:24:19 2018 kern.warn kernel: [   13.174702] ath10k_pci 0000:00:00.0: Direct firmware load for ath10k/QCA9887/hw1.0/board-2.bin failed with error -2
Mon Nov  5 21:24:19 2018 kern.warn kernel: [   13.185539] ath10k_pci 0000:00:00.0: Falling back to user helper
Mon Nov  5 21:24:19 2018 kern.err kernel: [   13.308055] firmware ath10k!QCA9887!hw1.0!board-2.bin: firmware_loading_store: map pages failed
Mon Nov  5 21:24:19 2018 kern.info kernel: [   13.325401] ath10k_pci 0000:00:00.0: board_file api 1 bmi_id N/A crc32 546cca0d
Mon Nov  5 21:24:19 2018 kern.err kernel: [   13.354858] ath10k_pci 0000:00:00.0: failed to read device EEPROM at 0000
Mon Nov  5 21:24:19 2018 kern.warn kernel: [   13.361892] ath10k_pci 0000:00:00.0: failed to read calibration data from EEPROM: -22
Mon Nov  5 21:24:19 2018 kern.info kernel: [   14.478574] ath10k_pci 0000:00:00.0: htt-ver 2.1 wmi-op 5 htt-op 2 cal otp max-sta 128 raw 0 hwcrypto 1
Mon Nov  5 21:24:19 2018 kern.debug kernel: [   14.616730] ath: EEPROM regdomain: 0x0
Mon Nov  5 21:24:19 2018 kern.debug kernel: [   14.616740] ath: EEPROM indicates default country code should be used
Mon Nov  5 21:24:19 2018 kern.debug kernel: [   14.616744] ath: doing EEPROM country->regdmn map search
Mon Nov  5 21:24:19 2018 kern.debug kernel: [   14.616759] ath: country maps to regdmn code: 0x3a
Mon Nov  5 21:24:19 2018 kern.debug kernel: [   14.616766] ath: Country alpha2 being used: US
Mon Nov  5 21:24:19 2018 kern.debug kernel: [   14.616770] ath: Regpair used: 0x3a
Mon Nov  5 21:24:19 2018 kern.debug kernel: [   14.787862] ath: EEPROM regdomain: 0x0
Mon Nov  5 21:24:19 2018 kern.debug kernel: [   14.787876] ath: EEPROM indicates default country code should be used
Mon Nov  5 21:24:19 2018 kern.debug kernel: [   14.787880] ath: doing EEPROM country->regdmn map search
Mon Nov  5 21:24:19 2018 kern.debug kernel: [   14.787900] ath: country maps to regdmn code: 0x3a
Mon Nov  5 21:24:19 2018 kern.debug kernel: [   14.787906] ath: Country alpha2 being used: US
Mon Nov  5 21:24:19 2018 kern.debug kernel: [   14.787910] ath: Regpair used: 0x3a

```

danielg4 commented:
> So, everything works good now. Just the EEPROM missing mac problem in some devices, but I think this is not an OpenWRT issue. Just in case it might be useful, the logread information regarding ath10k driver.

It most likely *is* an OpenWrt issue. Many ath10k devices have calibration partitions which must be manually loaded through scripts.

p4u commented:
> It most likely _is_ an OpenWrt issue. Many ath10k devices have calibration partitions which must be manually loaded through scripts.

You probably right

```
[   42.604526] ath10k_pci 0000:00:00.0: unable to read from the device (-145)
[   42.611652] ath10k_pci 0000:00:00.0: could not read board ext data addr (-145)
[   42.619153] ath10k_pci 0000:00:00.0: could not push board ext data (-145)
[   42.626183] ath10k_pci 0000:00:00.0: failed to download calibration data from EEPROM: -145
```

dangowrt commented:
Regarding the ath10k eeprom data: please retry with my updated staging tree which now tries to grab the eeprom data from flash just it works on other oolite devices. If that works we are ready for merge.

dangowrt commented:
Otherwise please try editing `target/linux/ar71xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata` until we get it to work.

p4u commented:
> Regarding the ath10k eeprom data: please retry with my updated staging tree which now tries to grab the eeprom data from flash just it works on other oolite devices. If that works we are ready for merge.

Confirmed, it works :) Great work @dangowrt
